google_guest_agent/instance_setup.go

```diff
@@ -195,6 +219,341 @@ func agentInit(ctx context.Context) {
+		smpAffinity := irq + "/smp_affinity_list"
+		if !isFile(smpAffinity) {
+			continue
+		}
+		virtionetIntxDir := irq + "/" + dev
+		virtionetMsixDirRegex := ".*/" + dev + "-(input|output)\\.([0-9]+)$"
+		if isDir(virtionetIntxDir) {
+			// All virtionet intx IRQs are delivered to CPU 0
+			logger.Infof("Setting %s to 01 for device %s.", smpAffinity, dev)
+			if err := ioutil.WriteFile(smpAffinity, []byte("01"), 0644); err != nil {
+				logger.Warningf("Could not write 01 to %s for %s.", smpAffinity, dev)
+			}
+			continue
+		}
+		// Not virtionet intx, probe for MSI-X
+		var virtionetMsixFound bool
+		irqDevs, err := filepath.Glob(irq + "/" + dev + "*")
+		if err != nil {
+			return err
+		}
+		var queueNum int
+		for _, entry := range irqDevs {
+			r, _ := regexp.Compile(virtionetMsixDirRegex)
+			if r.MatchString(entry) {
+				virtionetMsixFound = true
+				queueNum, err = strconv.Atoi(r.FindStringSubmatch(entry)[2])
+				if err != nil {
+					return err
+				}
+			}
+		}
+		affinityHint := irq + "/affinity_hint"
+		if !virtionetMsixFound || !isFile(affinityHint) {
+			continue
+		}
+
+		//Set the IRQ CPU affinity to the virtionet-initialized affinity hint
+		logger.Infof("Setting %s to %d for device %s.", smpAffinity, queueNum, dev)
+		if err = ioutil.WriteFile(smpAffinity, []byte(strconv.Itoa(queueNum)), 0644); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func configNVME(totalCPUs int) error {
+	var currentCPU = 0
+	devices, err := filepath.Glob(nvmeDevice)
+	if err != nil {
+		return err
+	}
+	for _, dev := range devices {
+		if !isDir(dev) {
+			continue
+		}
+		irqs, err := filepath.Glob(dev + "/msi_irqs/*")
+		if err != nil {
+			return err
+		}
+
+		for _, irqInfo := range irqs {
+			if !isFile(irqInfo) {
+				continue
+			}
+			currentCPU := currentCPU % totalCPUs
+			cpuMask := 1 << currentCPU
+			irq := path.Base(irqInfo)
+			logger.Infof("Setting IRQ %s smp_affinity to %d", irq, cpuMask)
+			if err := ioutil.WriteFile("/proc/irq/"+irq+"/smp_affinity", []byte(strconv.Itoa(cpuMask)), 0644); err != nil {
+				return err
+			}
+			currentCPU++
+		}
+	}
+	return nil
+}
+
+func configSCSI(totalCPUs int) error {
+	var irqs []int
+	devices, err := filepath.Glob(scsiDevice)
+	if err != nil {
+		return err
+	}
+	for _, device := range devices {
+		var ssd = 0
+		targetPaths, err := filepath.Glob(device + "/host*/target*/*")
+		if err != nil {
+			return err
+		}
+		for _, targetPath := range targetPaths {
+			if !isFile(targetPath + "/model") {
+				continue
+			}
+			b, err := ioutil.ReadFile(targetPath + "/model")
+			if err != nil {
+				return err
+			}
+			match, err := regexp.MatchString(".*EphemeralDisk.* ", string(b))
+			if err != nil {
+				return err
+			}
+			if match {
+				ssd = 1
+				queuePaths, err := filepath.Glob(targetPath + "/block/sd*/queue")
+				if err != nil {
+					return err
+				}
+				for _, queuePath := range queuePaths {
+					ioutil.WriteFile(queuePath+"/scheduler", []byte("noop"), 0644)
```

Comment:
have you tested this btw? `echo whatever > file` in bash will include a newline, but this WriteFile call won't, unsure if that will matter

Reply:
add a new line.